src/Lean/Elab/PreDefinition/PartialFixpoint/Induction.lean

```diff
@@ -66,90 +83,152 @@ private def numberNames (n : Nat) (base : String) : Array Name :=
-            let e' ← mkLambdaFVars adms e'
-            let e' ← mkLambdaFVars motives e'
-            let e' ← mkLambdaFVars (binderInfoForMVars := .default) (usedOnly := true) xs e'
-            let e' ← instantiateMVars e'
-            trace[Elab.definition.partialFixpoint.induction] "complete body of fixpoint induction principle:{indentExpr e'}"
-            pure e'
+    let e' ← if eqnInfo.fixpointType.all isLatticeTheoretic then
+      let leastOrGreatest : Option Bool := match eqnInfo.fixpointType[0]! with
+        | .leastFixpoint => Option.some true
+        | .greatestFixpoint => Option.some false
+        | _ => Option.none
+      if (eqnInfo.declNames.size != 1) then
+        throwError "Mutual lattice (co)induction is not supported yet"
+      eqnInfo.fixedParamPerms.perms[0]!.forallTelescope infos[0]!.type fun xs => do
+        -- Now look at the body of the functions
+        let body ← eqnInfo.fixedParamPerms.perms[0]!.instantiateLambda infos[0]!.value xs
+        let body' := PProdN.stripProjs body.eta
+        -- We strip it until we reach an application of `lfp_montotone`
+        let some fixApp ← whnfUntil body' ``lfp_monotone
+          | throwError "Unexpected function body {body}, could not whnfUntil lfp_monotone"
+        let_expr lfp_monotone α instcomplete_lattice F hmono := fixApp
+          | throwError "Unexpected function body {body}, not an application of lfp_induction"
+        let e' ← mkAppOptM ``lfp_induction_monotone #[α, instcomplete_lattice, F, hmono]
+        -- We get the type of the induction principle
+        let eTyp ← inferType e'
+        let f ← mkConstWithLevelParams infos[0]!.name
+        let fEtaExpanded ← lambdaTelescope infos[0]!.value fun ys _ =>
+            mkLambdaFVars ys (mkAppN f ys)
+        let fInst ← eqnInfo.fixedParamPerms.perms[0]!.instantiateLambda fEtaExpanded xs
+        let fInst := fInst.eta
+
+        -- Then, we change the conclusion so it doesn't mention the `lfp_monotone`, but rather the actual predicate.
+        let newTyp ← forallTelescope eTyp (fun args econc =>
+          if (econc.isAppOfArity ``PartialOrder.rel 4) then
+          let oldArgs := econc.getAppArgs
+          let newArgs := oldArgs.set! 2 fInst
+          let newBody := mkAppN econc.getAppFn newArgs
+          mkForallFVars args newBody
+          else
+            throwError "Unexpected conclusion of the fixpoint induction principle: {econc}"
+        )
+        -- Desugar partial order on predicates in premises and conclusion
+        let newTyp ← forallTelescope newTyp (fun args conclusion => do
+          let predicate := args[0]!
+          let predicateType ← inferType predicate
+          let premise := args[1]!
+          let premiseType ← inferType premise
+          let premiseType ← desugarOrder predicateType premiseType (reduceRhs := true) leastOrGreatest
+          let newConclusion ← desugarOrder predicateType conclusion
+            (fixpointType := leastOrGreatest)
```

Comment:
Inconsistent use of explicit argument name or not.

Maybe leave a comment why you `reduceRhs` in one call, but not the other?